### ISSUE TITLE
stream: write small messages cause massive memory usage

### ISSUE THREAD

**Issue by trevnorris** [nodejs/node-v0.x-archive#6623]
If I write until `writable.write()` returns false when writing many small buffers Node consumes a lot of memory and the process seems to halt:

``` javascript
var net = require('net');

net.createServer(function(conn) {
  conn.on('drain', function() {
    write(conn);
  });
  write(conn);
}).listen(8000, function() {
  console.log('Listening on port 8000');
});

var SIZE = 1024;
var b = new Buffer(SIZE);
var cntr = 0;

b.fill('a');

function write(conn) {
  while(conn.write(b)) {
    cntr += SIZE;
  }
}

var t = process.hrtime();
setInterval(function() {
  t = process.hrtime(t);
  var mib = cntr * 8 / 1024 / 1024 / (t[0] + t[1] / 1e9);
  console.log(mib.toFixed(2) + ' Mb/sec');
  cntr = 0;
  t = process.hrtime();
}, 2000);
```

As soon as the connection is made the `setInterval()` stops printing and memory begins to fill up quickly.

Here's another example that definitely shouldn't be able to fill up memory:

``` javascript
var net = require('net');

var SIZE = 1024;
var b = new Buffer(SIZE);
var cntr = 0;

b.fill('a');

net.createServer(function(conn) {
  (function write() {
    // Notice here we're passing write() as the flush callback.
    conn.write(b, write);
    cntr += SIZE;
  }());
}).listen(8000, function() {
  console.log('Listening on port 8000');
});

var t = process.hrtime();
setInterval(function() {
  t = process.hrtime(t);
  var mib = cntr * 8 / 1024 / 1024 / (t[0] + t[1] / 1e9);
  console.log(mib.toFixed(2) + ' Mb/sec');
  cntr = 0;
  t = process.hrtime();
}, 2000);
```

I run these tests by starting up the Node process then running `nc localhost 8000 > /dev/null`.

The second one doesn't cause the same memory issue in v0.10, while the first one does.

**Comment by dipbhi** [comment]
Is there any fix planned for this issue? I am still encountering it in v. 0.12.2.

**Comment by trevnorris** [comment]
@dipbhi I haven't taken another look at this since it my last post here. The quickest solution for you would be to occasionally use `setImmediate()` before continuing with the write. I'll put this back on my list of things to investigate.

**Comment by dipbhi** [comment]
Ok.  I will try setImmediate. Thanks for the suggestion.

**Comment by dipbhi** [comment]
By the way, would it be bad to call setImmediate before each write call?

**Comment by ashelley** [comment]
@dipbhi if you can listen for the write to return false then wait for the drain event to be emitted you probably won't see as much memory usage... 

pseudo code:
```
if(!stream.write(chunk) {
      // stop writing
     stream.once('drain', function() {
          //continue writing
     });
}
```

alternatively you can see watch stream._writableState.needDrain then wait for the drain event.  This will be more efficient than calling setImmediate on each write AFAIK.

**Comment by dipbhi** [comment]
@ashelley,
Nice suggestions, I will try them out. Thank you

**Comment by ashelley** [comment]
Actually, after reviewing this again the original example of using the network stream the highWaterMark is never hit as originally reported.  So the solution might depend on the type of stream you are working with.  

As @trevnorris mentioned setImmediate is your only option if you are using the socket stream.  One thing I did notice though that instead of monitoring stream._writableState.needDrain you can monitor stream._writableState.pendingcb to throttle your writes:

```
var net = require('net');

net.createServer(function(conn) {
  conn.on('drain', function() {
    write(conn);
  });
  write(conn);
}).listen(8000, function() {
});

var SIZE = 1024;
var b = new Buffer(SIZE);
var cntr = 0;

b.fill('a');

function write(conn) {
  while(conn.write(b)) {
    if((conn._writableState.pendingcb * SIZE) > conn._writableState.highWaterMark) {
      return setImmediate(function() {
        write(conn);
      });
    }
    cntr += SIZE;
  } 
}

var t = process.hrtime();
setInterval(function() {
  t = process.hrtime(t);
  var mib = cntr * 8 / 1024 / 1024 / (t[0] + t[1] / 1e9);
  console.log(mib.toFixed(2) + ' Mb/sec');
  cntr = 0;
  t = process.hrtime();
}, 2000);

```

**Comment by trevnorris** [comment]
@ashelley Great points. The addition to reducing the number of calls to `setImmediate()` is a great tip on improving performance.

**Comment by dipbhi** [comment]
@ashelley & @trevnorris 
Using setImmediate seems to be an excellent solution to control memory usage with one client server pair, and @ashelley's throttling suggestion has a throttling advantage.

However when I try to use similar solution with multiple clients (e.g. chat room) memory usage still goes up and server does seem to have a memory leak as well. I am hoping that I can pull your brain a little more to see if there is any workaround for multi-client situation as well.

Here is my sample code (server should be run with --expose-gc):

**Server**:
```javascript
var net = require('net'), conns = [];

net.createServer(function(conn) {
	console.log('client connected');
	conns.push(conn);

	conn.on('data', function(buf){
		distribute(conn, buf);
	});
	
	conn.on('error', function(err){
		//console.error(err);
		removeConn(conn);
	})

	conn.on('end', function(){
		console.log('client disconnected');
		removeConn(conn);
	});
  
}).listen(8030);

var SIZE = 1024;
var b = new Buffer(SIZE);
var cntr = 0;
b.fill('a');

function distribute(conn, buf) {
	for(var i = 0, len = conns.length; i < len; i++){
		var targetConn = conns[i];
		if(targetConn != conn){
			write(targetConn, buf);	
		}
	} 
}

function write(conn, buf) {
	if(conn){
		if(conn._writableState.pendingcb > conn._writableState.highWaterMark) {
			return setImmediate(function() {
				write(conn, buf);
			});
		}
		else{
			conn.write(buf, function(err){
				if(err){
					removeConn(conn);
				}
			});
		}
	}
    cntr += SIZE;
}

function removeConn(conn){
	var connIndex = conns.indexOf(conn);
	if(connIndex > 0){
		conns.splice(connIndex, 1);
	}	
}

var t = process.hrtime();
setInterval(function() {
  t = process.hrtime(t);
  var mib = cntr * 8 / 1024 / 1024 / (t[0] + t[1] / 1e9);
  console.log(mib.toFixed(2) + ' Mb/sec');
  cntr = 0;
  t = process.hrtime();
}, 2000);

setInterval(function(){
	console.log('calling gc');
	gc();
}, 30000);
```

**Client**:
```javascript
var cp = require('child_process'),
	numOfChildren = 100,
	runningChildren = numOfChildren;	
	
for(var i = 0; i < numOfChildren; i++){
	var child = cp.fork('child.js', {silent: true});

	child.on('close', function(){
		runningChildren--;
		if(runningChildren == 0){
			console.log('Execution completed.');
		}
	});
	
	child.stdout.on('data', function(data){
		console.log('' + data);
	});
	
	child.stderr.on('data', function(data){
		console.error(data.toString());
	});		

}
```

**Client Child**
```javascript
var net = require('net'), 
	numOfMsgs = 1000, 
	receivedCount = 0, 
	receivedPrintFreq = 1000,
	host = '127.0.0.1',
	totalBufferSize = 0, 
	writeCount = 0;

var conn = net.connect({port: 8030, host: host}, function() {
	console.log('Connected to the server, pid - %d', process.pid);
});

conn.on('data', function(buffer){
	receivedCount++;
	totalBufferSize += buffer.length;
	if(receivedCount % receivedPrintFreq === 0) console.log('%d messages received, buffer size - %d, pid - %d', receivedCount, totalBufferSize, process.pid);
});

conn.on('error', function(err) {
  console.error(err);
});

conn.on('end', function() {
  console.log('Disconnected from the server, pid - %d', process.pid);
});

var SIZE = 1024;
var b = new Buffer(SIZE);

b.fill('a');

for(var i = 0; i < numOfMsgs; i++){
	conn.write(b, function(){
		writeCount++;
		if(writeCount === numOfMsgs){
			setImmediate(function(){
				conn.end();	
			});
		}
	});
}
```

**Comment by saquibkhan** [comment]
Can you try on node v0.12.3 ?  I tried the first test and i didn't see any memory leaks.

**Comment by dipbhi** [comment]
@saquibkhan  I was using latest release v0.12.2. Is there any installer available for v0.12.3?

**Comment by saquibkhan** [comment]
hmm, v0.12.2 should also be fine. i used below code to test me leaks, i could see me leak on v0.10 but not on v0.12 dev branch
```javascript
var net = require('net');

net.createServer(function(conn) {
  conn.on('drain', function() {
    write(conn);
  });
  write(conn);
}).listen(8000, function() {
  console.log('Listening on port 8000');
});

var SIZE = 1024;
var b = new Buffer(SIZE);
var cntr = 0;

b.fill('a');

function write(conn) {
  while(conn.write(b)) {
    cntr += SIZE;
  }
}

var t = process.hrtime();
setInterval(function() {
  t = process.hrtime(t);
  var mib = cntr * 8 / 1024 / 1024 / (t[0] + t[1] / 1e9);
  console.log(mib.toFixed(2) + ' Mb/sec');
  cntr = 0;
  t = process.hrtime();
}, 2000);
```

**Comment by dipbhi** [comment]
@saquibkhan OK. How about the multi-client scenario I posted later. Does that leak on the dev branch?

**Comment by jasnell** [comment]
@trevnorris ... what do you want to do with this one?

**Comment by trevnorris** [comment]
@jasnell This is an issue because we call the callback sync if all data was flushed immediately. Which prevents the loop from turning over, thus not allowing full cleanup.

We could do a simple patch to force async after N calls. Same thing that libuv does. After 49,000 events are processed in the poll phase it forces the event loop to proceed. Should be a simple patch, but never got around to it.

**Comment by d3x0r** [comment]
My issue is more along the lines of https://github.com/nodejs/node-v0.x-archive/issues/8048 but they say that that is a dup of this.

```
"use strict";

var fs = require( 'fs' );
var wstream;
var fileNum = 1;
var basePath = "./";
var baseName = "output_file";
var maxlimit = 100000000;
var limit = 1000000;
var subCount = 0;
var cardGen = 0;
var single_buffer = new Buffer( 36 );

function doGeneration() {
    var run = true;
        // done generating
        if( cardGen >= maxlimit )
            return;
       cardGen += limit;
    if( !wstream ) {
      console.log( `create streams... ${fileNum}`);
      wstream = fs.createWriteStream(  basePath + baseName + `.${fileNum}.dat` );
      wstream.on( "close", ()=>{ console.log( "Fail closed." );
      setTimeout( ()=>{ global.gc(); doGeneration() }, 10 );
        } );
      wstream.on( "end", ()=>{ console.log( "file ended." );} );
      fileNum++;
    }
    for(subCount = 0; subCount < limit; subCount++ ) {
        wstream.write( single_buffer );
    }
    {
      console.log( "end stream?" );
      wstream.end();
      wstream = null;
    }
}

doGeneration();


```

That sample even uses a setTimeout after the file closes... so everything should be flushed and done before the next iteration of this process.
My Original process I noticed the problem when I didn't use setTimeout(); but when I would, then I would get huge delays between runs when nothing was happening (no disk, no memory)... this stripped sample slowly runs out of memory even with the global.gc(); on every file close... takes a couple minutes; generates a few hundred megs of data.

node --expose-gc --max-old-space-size=512 test.js >err 2>&1
(hmm... still used over 1.6G even with smaller limit )

but unfortunately that doesn't really run out of memory like my actual process though... but making it might give me a work around to wait until the close event to dispatch another run.

**Comment by ashelley** [comment]
@d3x0r if you listen for the return value of write and wait for the drain event to happen on wstream before continuing you probably won't see the problem. in other words... you can't write in a for loop without listening to the return value of write.  when write returns false you have to wstream.once('drain', () => {//start running loop again})

**Comment by d3x0r** [comment]
@ashelly thanx; I will attempt that monday... other fish to shish-kabob.  I see there's often a 'right' way that's the only way.  I'm a little astounded that the file can't just evaporate with all it's buffers... even if it takes a 'while' ... at least scheduling in the `on("close",()=>{ global.gc(); doGenerate() } ); /* generator.next()? */` it's hovering at 1.2G more than halfway done.  
I still don't know which direction to go to simulate the stall condition though; it's not the first time that I did setTimeout(f,0) and has the program just 'stop' but using setTimeout(f, 8); worked OK.  and that is in a serial process that does a bit of work on a single clause of data and then posts the results and reschedules to get another.  The read/write is actually in in the cluster.master. but I guess that could use once( 'drain' )  /\* ?once? nice. */

```
"use strict";

var fs = require( 'fs' );
var fileNum = 1;
var basePath = "./";
var baseName = "output_file";
var maxlimit = 100000000;
var limit = 1000000;
var cardGen = 0;
var single_buffer = new Buffer( 36 );

function doGeneration() {
   // for( generation = 0; gneration < 1000; generation++ ) { 
    var wstream;
    if( cardGen >= maxlimit )   return;
    if( !wstream ) {
      //console.log( `create streams... ${fileNum}`);
      wstream = fs.createWriteStream(  basePath + baseName + `.${fileNum}.dat` );
      fileNum++;
    }
    for(var subCount = 0; subCount < limit; subCount++ )  wstream.write( single_buffer );
    cardGen += subCount;
    wstream.end();
    wtream - null;
   //}
    return doGeneration();
}

setTimeout( doGeneration );

```

**Comment by d3x0r** [comment]
I updated my comment to include the sample I'd rather do, without all the bells and whistles of any events and possible reverse-reference linking...

---

I see, and the next logical step is after implementing on( 'drain', ()=>{} ) for output write failures, if the input is readline file stream, the lines don't stop immediately... but it could be possible to end() the file before the drain fires.  in that case the drain never fires, and never unlocks the thing... so you have to keep the closed 

This version shouldn't run out of memory.... and yet it does.  Hooray for garbage collection (NOT!)

```
"use strict";

var fs = require( 'fs' );
var fileNum = 1;
var basePath = "./";
var baseName = "output_file";
var maxlimit = 100000000;
var limit = 1000000;
var subCount = 0;
var cardGen = 0;
var single_buffer = new Buffer( 36 );
var blocked = false; // external things use this to see if they can continue...
var wstream = null;;
var wstreamClosed = null;;

function doGeneration() {
    if( cardGen >= maxlimit )   return;
    if( !wstream ) {
      wstream = fs.createWriteStream(  basePath + baseName + `.${fileNum}.dat` );
      wstream.on( 'drain', () =>{ 
         if( wstreamClosed ) {
           wstreamClosed.end();
           wstreamClosed = null;
        }
        blocked = false;
        setTimeout( doGeneration ); // continue generation...
      } );
      fileNum++;
    }
    for(  ; !blocked && subCount < limit; subCount++ )  
        if( !wstream.write( single_buffer ) )
       {
            blocked = true;
            return;
       }
    if( !blocked )
    {
        cardGen += subCount;
        subCount = 0;
        if( blocked )  wstreamClosed = wstream;
        else       wstream.end();
        wstream = null;
        setTimeout( doGeneration )
        //return doGeneration();
    }
}

setTimeout( doGeneration );
```

**Comment by d3x0r** [comment]
@ashelley  I updated the sample to include drain event listening and protection to avoid end()ing the file... (see prior post)  

So even doing it the 'proper' way doesn't fix leaks.  
by 45000000 it's at 700MB.

**Comment by ashelley** [comment]
i'm not sure how to limit 'new node''s memory usage but this code works fine and completes successfully in a small memory footprint.  I ran your above code sans es6 syntax in node 0.12.7 and it works fine: node --max_old_space_size=20 test.js

I doubt that --max_old_space_size works on new versions of node as it probably allocates memory differently and requires a different command line switch to "properly limit it".  

The key is that once you are listening to drain you don't go into "mystery wait times" and everything performs as expected.

I'd look for answers regarding on how to limit node's memory usage from command line options and you will see that with your above code you will get your expected results.

**Comment by d3x0r** [comment]
(The sample code as is, as of this time 
it doesn't apparently limit the memory... 
It does allow increase above 1.5G for 64bit node... 5120 for instance works.... 

the code in a nightly build from a few weeks ago does generate large files...
but it might be disk performance vs cpu performance... if you have a slower cpu maybe you won't get disk bottle necked... was worried it was something timing sensitive since the GC works in a schedule of time maybe it's starved?  I didn't addditionally apply manual gc to it to see.

When I first was diagnosing this, I used electron which uses node, and used its debugging tools...but when I paused the code so I could do a heap snapshot the heap cleaned itself and went back to 340M (typical for tables kept in memory) ... from 1G.. 700M just because of writing to a file is obscene.

core i7-6700 3.4ghz
ST1000D1M003-1ER162  (probably http://www.newegg.com/Product/Product.aspx?Item=N82E16822148840 )
node V5.8.0

---

And it's not even like this is a threading problem.  I'm definitely providing a time on the stack when there is 0 program events.   Collision with doing end() while it is in the state of needing to emit drain()?

I did try with once, but the input was a readline thing, and .pause() does... after 20 some messages.  And setting up a .once handler blinding not checking to see if I had one set already caused a over-scheduling warning from addEventListener.

It just gets more complex when the linereader feeds a cluster that has work queues that come out and feed the output to a file... it's like 'oh by the way, when the cluster gives you an event, don't do anything with it because I have to wait for a drain event'

---

This is why people say 'use a real streaming library like XXXXX' that has 20+ other dependancies itself.... I could rework this as transform filters maybe that wouldn't leak as much?

**Comment by ashelley** [comment]
internally there is an array that gets sliced 1 time per write. so if you don't wait for drain the internal buffer grows huge... to reproduce the issue create a large array and remove each element by slicing it from the bottom.  it's _slow_.  by listenning to the drain event you cause the internal buffer to not grow huge so you don't end up eventually waiting for an array to get sliced inefficiently once per write.   it's definitely a bug in node not being able to flush writes out in batches and inefficiently handling the internal buffer.

**Comment by d3x0r** [comment]
Okay... so let's suppose wstream is an array that gets things added, and slices a bit off every time and keeps adding things... the drain event supposedly when wstream.length == 0 

Edit3: _I lied_ it's losing about 4k per second after running a half hour or so... and it's up to 42M...

In ideal conditions this appears to go to 39M and stay there.  It started at 22M or so... 
I don't know if I would have to tune the constants to make better pressure on that sort of thing....

I suppose I could copy the bytes flat into the buffer... that would be more like write.

```
"use strict";

var fileNum = 1;
var maxlimit = 100000000000000;
var limit = 1000000;
var subCount = 0;
var cardGen = 0;
var single_buffer = new Buffer( 36 );
var wstream = null;;

function doGeneration() {
    if( cardGen >= maxlimit )   return;
    if( !wstream ) {
      wstream = [];

   }
    for(  ; subCount < limit; subCount++ )  
    {
        wstream.push( single_buffer/*.clone() ? */ );
        if( ( subCount % 1000 ) == 0 ) {
            wstream = wstream.slice( -Math.floor( wstream.length *0.10 ) );
            //console.log( `slice is... ${wstream.length}` ); // goes from 100 to 111 and stays
        }
    }

    subCount = 0;
    cardGen += limit;    
    //console.log( `stream is ${wstream.length}` );  /// finally flush everything eventually and relase it?
    // actually this starts another buffer.. probably the 111 is kept around forever because the new file
    // will have priority over the old file even though the old file does eventually drain and close? (does it?)
    wstream = null;
    setTimeout( doGeneration )
    //return doGeneration();
}

setTimeout( doGeneration );
```

**Comment by ashelley** [comment]
the code has changed since i last looked but you are better off trying to debug the actual node code i'd say instead of trying to mock it:

https://github.com/nodejs/node/blob/v4.4.2-proposal/lib/_stream_writable.js#L398-L412

EDIT: i think it was stated above that it is because this cb never called async but i could be wrong.

https://github.com/nodejs/node/blob/v4.4.2-proposal/lib/_stream_writable.js#L355

EDIT2: I actually added global.gc() back to your code and it started garbage collecting keeping the memory at around 300-400MB (err maybe not)

EDIT3: after playing around with it a bit... node 0.12 doesn't go crazy with memory usage with this code but writes a lot slower (more cpu i think too).  4.4.2 and 5.10.1 write a lot faster but eat memory so i believe the issue here is totally different depending on which version of node you are using.

**Comment by d3x0r** [comment]
Still broken in 6.1.0 _bump_ (or is bumping bannable?)
1) that sort of project is done now and I have no need to fix it myself? (I am inclined to be a nuisance every few months as new versions come out and the test still fails)
2) others use external libraries for streaming that do better jobs?
3) is it a core v8 issue? Is this a dup of their bug?  Because I played with doing some file system objects and I had to inject a 1.4 second wait before global.gc would gc.  It's as simple as having a destructor defined, and waiting patently. ( a file system may have multiple files in itself too which may further complicate the waiting)... and I can see starving it like this is a sin; (fixable with appropriate options and features; but couldn't the native parts somehow kick the garbage collector for me?)
